### PR TITLE
RFP-10 Milestone 2

### PR DESCRIPTION
**Pull request by chappjc**
Milestone 2 of RFP-10 includes the following legacy RPC tests:

- [x] sendtoaddress
- [x] sendfrom
- [x] sendmany
- [x] listtransactions
- [x] getwalletfee/settxfee
- [x] getticketfee/setticketfee

Note that transaction fee (aka relay or wallet fee) and ticket fee gets are tested via walletinfo since there are no more direct methods available via dcrrpcclient.  Wallet.RelayFee() is the shared underlying call.  The milestone 3 test, purchaseticket, is also completed.

Final tasks:
- [x] Review test procedures
- [x] Improve documentation
- [x] Provide summary for squash merge.

This PR also includes the addition of the mkharness command (dcrwallet/rpctest/mkharness/main.go), which brings up a simnet node and wallet configured for mining.  Once set up, mkharness will print the commands for the running node and wallet, and suggested dcrctl command constructions.

Some questions and possible TODO:
 * In comparing nominal vs actual/observed fees, an arbitrary tolerance is used.  absFeeDiffTolerance = 0.0005  Is this appropriate?
 * Is it correct that ListUnspent should only return coinbase tx that are validated (>=2 confirmations)?
 * Is it expected that GetRawTransaction (and similar) will have incorrect values, at least in vin, until the transaction is mined and in the caes of sstx, validated?  Generating a block before getrawtransaction is needed to compute fee.
 * For milestone 3: doc.go

### CODE REVIEW

**Comment by chappjc** [comment]
For squash merge:

This completes RFP-10 milestone 2.

Tests added or completed:

sendtoaddress, sendfrom, sendmany, listtransactions, getwalletfee/settxfee, getticketfee/setticketfee, purchasetickets (milestone 3)

Other fixes/updates included:
 * Create mkharness utility for starting up a simnet dcrd and dcrwalet, and necessary edits to rpctest harness system.
 * Add `--txindex` and `--addrindex` flags to node command to work with dcrd updates related to new database package.
 * Create multi-harness system allowing  a test to use a dedicated harness when specified, otherwise a common harness is used.
 * Add getOutPointString() utility function to create an OutPoint from a ListUnspentResult and call it's String() method.
 * Add includesTx(), which checks if a block contains a transaction hash.
 * Add getWireMsgTxFee(), which computes the effective absolute fee from a Tx.
 * Plugin a OnBlockConnected handler to help verify intended function and track generation in tests.
 * Remove redundant RPCClient, h.chainClient.
 * Tweaks to timing and logging in harness setup.
 * Add .gitignore just for rpctest.

**Comment by alexlyp** [comment]
@chappjc  Looking good so far, will make any comments I have in the code.

Re: your 3 questions above:
1) They should be int64s for fees so you should need to have a tolerance, right?
2) Let me confirm with @cjepson but I believe that is the case.  Though I think it could be until they mature (CoinbaseMaturity in chaincfg/params.go)
3) Yes there are a few fields that will not be populated there until the tx is mined.  So typically I've waited until the tx is mined to check the fee on it.  Not ideal, but no other way to do it without adding a "estimate tx size" utility or something

File: rpctest/rpcharness_test.go [hunk 1]
```diff
@@ -972,31 +1064,736 @@ func testSendFrom(r *Harness, t *testing.T) {
...
+	// Give the tx a sensible MsgTx().TxIn[:].ValueIn values
+
+	// Compute the fee
+	rawTx, err := r.Node.GetRawTransaction(txid)
+	if err != nil {
+		t.Fatalf("getrawtransaction failed: %v", err)
+	}
+
+	fee := getWireMsgTxFee(rawTx)
+	feeRate := fee.ToCoin() / float64(rawTx.MsgTx().SerializeSize()) * 1000
```
Comment (chappjc) [inline rpctest/rpcharness_test.go]: Say this was done in Amount, there would still be division by size, and there might be a remainder, essentially an inaccuracy.

**Comment by chappjc** [comment]
@alexlyp For the fee checks, it's not just that ToCoin() is used, but that computing the effective fee rate even in int64s requires a division, where there's no guarantee the remainder will be zero.  It seems not to be.
https://github.com/decred/dcrwallet/pull/336/files#diff-de52fd83fe5056396f8b126760ae5316R1543
Maybe I'm thinking about this wrong.  Maybe SerializeSize is not accurate?

**Comment by chappjc** [comment]
Regarding coinbase in listunspent, it does show confirmations=2, but what's unexpected is that even with ListUnspentMinMax(minconf=0|1,...) it doesn't show the unvalidated coinbase tx.

**Comment by alexlyp** [comment]
Hmm yah must be a bug with ListUnspent then.  For now, expecting it to show upon 2 conf is acceptable imo.

**Comment by alexlyp** [comment]
So after discussing with @jrick, all the fee/size estimation is tested very thoroughly in the unit tests. We think you should make sure that the observed fee is ABOVE the set fee.   So you can use whatever arbitrary amount you think is worthy.  We can always tweak it in the future if need be.

**Comment by chappjc** [comment]
Ha, I think I broke it by adding one too many new blocks before any tickets were purchased and it hit validation height without tickets to vote on it..  Geez, i though about this, but I thought there was time before the purchaseticket test was called.  Will have to fix this.

**Comment by alexlyp** [comment]
Hmmm passed for me just a bit ago, strange.

But yah voting height is 143 on simnet.  I typically let my simnet wallets by max tickets around ~100 and pool is able to fill up decently before then (since you have to take into account 16 block ticket maturity as well)

**Comment by chappjc** [comment]
I even made a note of this,
https://github.com/decred/dcrwallet/pull/336/files#diff-de52fd83fe5056396f8b126760ae5316R1788
Basically, tickets must be purchased before block 128 or the chain dies.

**Comment by chappjc** [comment]
@alexlyp Well, the fee thing is wrapped up, and it's working.  But I'm still not happy about how dcrwallet get sooo far behind the daemon in this loop:  https://github.com/chappjc/dcrwallet/blob/rfp10-m2/rpctest/rpcharness_test.go#L1793

It's not an issue with rpctest akaict, because the wallet is just super slow to connect blocks.  Interestingly (as I mentioned in IRC), even though the wallet votes when a block comes in, it gets way behind and the block connected messages will be minutes behind by block 100 or so.

**Comment by alexlyp** [comment]
@chappjc Sorry for the delay, some webwallet issues arose and needed squash them. 

Yah @jcvernaleo and I have fought with dcrwallet slowness when connecting blocks since the beginning of decred dev. 

And yes purchaseticket is painfully slow.  That has to do with the way in which inputs are selected I believe.  

Both of these issues are on our radar and hopefully as optimizations occur it will help future improvements to this rpctest harness.

**Comment by alexlyp** [comment]
OK looks great.  Ready to merge. 

Can you rebase/squash this and put a short summary of the changes you have here for the squashed commit message.  Here is the example I was able to do for your previous milestone: https://github.com/decred/dcrwallet/commit/9987b2960f9c725e88a8f5585215718b4447f56b

**Comment by chappjc** [comment]
@alexlyp No squash merge this time?  I drafted a commit message for this above https://github.com/decred/dcrwallet/pull/336#issuecomment-245505841

**Comment by alexlyp** [comment]
Ah I see your post above.  Exactly what I was looking for :)  Thanks.  Merging now.

**Pull request merged**